File: native/ControlUnit.cpp
// Copyright (c) 2026 Huawei Technologies Co., Ltd.
// SPDX-License-Identifier: CANN-1.0

#include "native/ControlUnit.h"

#include "native/ISATraits.h"

#include <algorithm>
#include <cctype>

namespace vfsim {
namespace {

std::string upper(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
    return static_cast<char>(std::toupper(c));
  });
  return value;
}

std::string normalizeBarrier(std::string value) {
  value = upper(std::move(value));
  const auto dot = value.rfind('.');
  if (dot != std::string::npos)
    value = value.substr(dot + 1);
  return value;
}

} // namespace

ControlUnit::ControlUnit(const ParamDB *db) : db_(db) {}

void ControlUnit::acceptMembar(const DynamicInst &inst) {
  const std::string barrier = normalizeBarrier(inst.barrier.empty() ? "VST_VLD" : inst.barrier);
  Barrier item;
  item.streamSeq = inst.streamSeq;
  item.pc = inst.pc;
  item.barrier = barrier;
  if (barrier == "VST_VLD") {
    item.waitClass = "STORE";
    item.blockClass = "LOAD";
  } else if (barrier == "VLD_VST") {
    item.waitClass = "LOAD";
    item.blockClass = "STORE";
  } else {
    if (db_ != nullptr) {
      db_->recordWarning(
          "unsupported_membar_type",
          {{"pc", std::to_string(inst.pc)},
           {"barrier", barrier.empty() ? inst.barrier : barrier},
           {"stream_seq", std::to_string(inst.streamSeq)}});
    }
    return;
  }
  barriers_.push_back(std::move(item));
}

void ControlUnit::update(
    const std::function<bool(int64_t, const std::string &)> &hasPendingBefore) {
  for (auto &barrier : barriers_) {
    if (barrier.released)
      continue;
    if (!hasPendingBefore(barrier.streamSeq, barrier.waitClass))
      barrier.released = true;
  }
  barriers_.erase(
      std::remove_if(barriers_.begin(), barriers_.end(),
                     [](const Barrier &barrier) { return barrier.released; }),
      barriers_.end());
}

bool ControlUnit::blocks(const DynamicInst &inst, const ParamDB &db,
                         const std::string &dtype) const {
  if (inst.type != "inst")
    return false;
  std::string cls;
  if (isLoadOp(db, inst.op, inst.form.empty() ? dtype : inst.form))
    cls = "LOAD";
  else if (isStoreOp(db, inst.op, inst.form.empty() ? dtype : inst.form))
    cls = "STORE";
  else
    return false;

  for (const auto &barrier : barriers_) {
    if (barrier.released)
      continue;
    if (inst.streamSeq > barrier.streamSeq && cls == barrier.blockClass)
      return true;
  }
  return false;
}

} // namespace vfsim

File: core/control_unit.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Dict, List

from core.isa_traits import get_op_class


@dataclass
class MembarState:
    stream_seq: int
    pc: int
    barrier: str
    wait_class: str
    block_class: str
    released: bool = False


class ControlUnit:
    """Control-side model for explicit vector memory barriers."""

    _SUPPORTED = {
        "VST_VLD": ("STORE", "LOAD"),
        "VLD_VST": ("LOAD", "STORE"),
    }

    def __init__(self, pdb: Any, dtype: str = "fp32") -> None:
        self.db = pdb
        self.dtype = str(dtype)
        self.barriers: List[MembarState] = []

    @staticmethod
    def normalize_barrier(raw: Any) -> str:
        text = str(raw or "VST_VLD").strip().upper()
        if "." in text:
            text = text.rsplit(".", 1)[-1]
        return text

    def accept_membar(self, node: Dict[str, Any]) -> None:
        barrier = self.normalize_barrier(node.get("barrier", node.get("type", "VST_VLD")))
        if barrier not in self._SUPPORTED:
            if hasattr(self.db, "record_warning"):
                self.db.record_warning(
                    "unsupported_membar_type",
                    barrier=barrier,
                    pc=int(node.get("pc", -1)),
                    stream_seq=int(node.get("stream_seq", -1)),
                )
            return
        wait_class, block_class = self._SUPPORTED[barrier]
        self.barriers.append(
            MembarState(
                stream_seq=int(node.get("stream_seq", -1)),
                pc=int(node.get("pc", -1)),
                barrier=barrier,
                wait_class=wait_class,
                block_class=block_class,
            )
        )

    def update(self, has_pending_prior: Callable[[int, str], bool]) -> None:
        for barrier in self.barriers:
            if barrier.released:
                continue
            barrier.released = not has_pending_prior(
                int(barrier.stream_seq),
                str(barrier.wait_class),
            )

    def blocks(self, inst: Dict[str, Any]) -> bool:
        stream_seq = int(inst.get("stream_seq", -1))
        if stream_seq < 0:
            return False
        op_class = get_op_class(
            inst.get("op", ""),
            self.db,
            str(inst.get("form", "") or self.dtype),
        )
        for barrier in self.barriers:
            if barrier.released:
                continue
            if stream_seq <= int(barrier.stream_seq):
                continue
            if op_class == barrier.block_class:
                return True
        return False
